File: src/main/java/com/refinedmods/refinedstorage/container/transfer/TransferManager.java
package com.refinedmods.refinedstorage.container.transfer;

import com.refinedmods.refinedstorage.apiimpl.API;
import com.refinedmods.refinedstorage.inventory.fluid.FluidInventory;
import net.minecraft.world.Container;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;
import net.neoforged.neoforge.items.IItemHandler;
import net.neoforged.neoforge.items.IItemHandlerModifiable;
import net.neoforged.neoforge.items.SlotItemHandler;
import javax.annotation.Nullable;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Supplier;

public class TransferManager {
    private final Map<IInventoryWrapper, List<IInventoryWrapper>> fromToMap = new HashMap<>();
    private final AbstractContainerMenu container;

    @Nullable
    private Function<Integer, ItemStack> notFoundHandler;

    public TransferManager(AbstractContainerMenu container) {
        this.container = container;
    }

    public void clearTransfers() {
        this.fromToMap.clear();
    }

    public void setNotFoundHandler(@Nullable Function<Integer, ItemStack> handler) {
        this.notFoundHandler = handler;
    }

    public void addTransfer(Container from, IItemHandler to) {
        addTransfer(new InventoryInventoryWrapper(from), new ItemHandlerInventoryWrapper(to));
    }

    public void addTransfer(Container from, Container to) {
        addTransfer(new InventoryInventoryWrapper(from), new InventoryInventoryWrapper(to));
    }

    public void addFilterTransfer(Container from, IItemHandlerModifiable itemTo, FluidInventory fluidTo, Supplier<Integer> typeGetter) {
        addTransfer(new InventoryInventoryWrapper(from), new FilterInventoryWrapper(itemTo, fluidTo, typeGetter));
    }

    public void addItemFilterTransfer(Container from, IItemHandlerModifiable to) {
        addTransfer(new InventoryInventoryWrapper(from), new ItemFilterInventoryWrapper(to));
    }

    public void addFluidFilterTransfer(Container from, FluidInventory to) {
        addTransfer(new InventoryInventoryWrapper(from), new FluidFilterInventoryWrapper(to));
    }

    public void addTransfer(IItemHandler from, Container to) {
        addTransfer(new ItemHandlerInventoryWrapper(from), new InventoryInventoryWrapper(to));
    }

    public void addBiTransfer(Container from, IItemHandler to) {
        addTransfer(from, to);
        addTransfer(to, from);
    }

    private void addTransfer(IInventoryWrapper from, IInventoryWrapper to) {
        List<IInventoryWrapper> toList = fromToMap.computeIfAbsent(from, k -> new LinkedList<>());

        toList.add(to);
    }

    public ItemStack transfer(int index) {
        Slot slot = container.getSlot(index);

        IInventoryWrapper key;
        if (slot instanceof SlotItemHandler) {
            key = new ItemHandlerInventoryWrapper(((SlotItemHandler) slot).getItemHandler());
        } else {
            key = new InventoryInventoryWrapper(slot.container);
        }

        List<IInventoryWrapper> toList = fromToMap.get(key);

        if (toList != null) {
            ItemStack initial = slot.getItem().copy();

            ItemStack remainder = slot.getItem();

            for (IInventoryWrapper to : toList) {
                InsertionResult result = to.insert(remainder);

                if (result.getType() == InsertionResultType.STOP) {
                    break;
                } else if (result.getType() == InsertionResultType.CONTINUE_IF_POSSIBLE) {
                    remainder = result.getValue();

                    if (remainder.isEmpty()) {
                        break;
                    }
                }
            }

            slot.set(remainder);
            slot.setChanged();

            if (API.instance().getComparer().isEqual(remainder, initial) && notFoundHandler != null) {
                return notFoundHandler.apply(index);
            }
        } else if (notFoundHandler != null) {
            return notFoundHandler.apply(index);
        }

        return ItemStack.EMPTY;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/container/transfer/TransferManager.kt
package com.refinedmods.refinedstorage.container.transfer

import com.refinedmods.refinedstorage.api.network.grid.IGrid.slotId
import com.refinedmods.refinedstorage.apiimpl.API.Companion.instance
import com.refinedmods.refinedstorage.inventory.fluid.FluidInventory
import net.minecraft.inventory.IInventory
import net.minecraft.inventory.container.Container
import net.minecraft.inventory.container.Slot
import net.minecraft.item.ItemStack
import net.minecraftforge.items.IItemHandler
import net.minecraftforge.items.IItemHandlerModifiable
import net.minecraftforge.items.SlotItemHandler
import java.util.*
import java.util.function.Function
import java.util.function.Supplier

class TransferManager(container: Container) {
    private val fromToMap: MutableMap<IInventoryWrapper, MutableList<IInventoryWrapper>> = HashMap()
    private val container: Container

    @Nullable
    private var notFoundHandler: Function<Int, ItemStack>? = null
    fun clearTransfers() {
        fromToMap.clear()
    }

    fun setNotFoundHandler(@Nullable handler: Function<Int, ItemStack>?) {
        notFoundHandler = handler
    }

    fun addTransfer(from: IInventory, to: IItemHandler?) {
        addTransfer(InventoryInventoryWrapper(from), ItemHandlerInventoryWrapper(to))
    }

    fun addTransfer(from: IInventory, to: IInventory) {
        addTransfer(InventoryInventoryWrapper(from), InventoryInventoryWrapper(to))
    }

    fun addFilterTransfer(from: IInventory, itemTo: IItemHandlerModifiable, fluidTo: FluidInventory, typeGetter: Supplier<Int>) {
        addTransfer(InventoryInventoryWrapper(from), FilterInventoryWrapper(itemTo, fluidTo, typeGetter))
    }

    fun addItemFilterTransfer(from: IInventory, to: IItemHandlerModifiable) {
        addTransfer(InventoryInventoryWrapper(from), ItemFilterInventoryWrapper(to))
    }

    fun addFluidFilterTransfer(from: IInventory, to: FluidInventory) {
        addTransfer(InventoryInventoryWrapper(from), FluidFilterInventoryWrapper(to))
    }

    fun addTransfer(from: IItemHandler?, to: IInventory) {
        addTransfer(ItemHandlerInventoryWrapper(from), InventoryInventoryWrapper(to))
    }

    fun addBiTransfer(from: IInventory?, to: IItemHandler?) {
        addTransfer(from, to)
        addTransfer(to, from)
    }

    private fun addTransfer(from: IInventoryWrapper, to: IInventoryWrapper) {
        val toList = fromToMap.computeIfAbsent(from) { k: IInventoryWrapper? -> LinkedList() }
        toList.add(to)
    }

    fun transfer(index: Int): ItemStack {
        val slot: Slot = container.getSlot(index)
        val key: IInventoryWrapper
        key = if (slot is SlotItemHandler) {
            ItemHandlerInventoryWrapper((slot as SlotItemHandler).getItemHandler())
        } else {
            InventoryInventoryWrapper(slot.inventory)
        }
        val toList: List<IInventoryWrapper>? = fromToMap[key]
        if (toList != null) {
            val initial: ItemStack = slot.getStack().copy()
            var remainder: ItemStack = slot.getStack()
            for (to in toList) {
                val result = to.insert(remainder)
                if (result.type == InsertionResultType.STOP) {
                    break
                } else if (result.type == InsertionResultType.CONTINUE_IF_POSSIBLE) {
                    remainder = result.value
                    if (remainder.isEmpty) {
                        break
                    }
                }
            }
            slot.putStack(remainder)
            slot.onSlotChanged()
            if (instance().getComparer()!!.isEqual(remainder, initial) && notFoundHandler != null) {
                return notFoundHandler!!.apply(index)
            }
        } else if (notFoundHandler != null) {
            return notFoundHandler!!.apply(index)
        }
        return ItemStack.EMPTY
    }

    init {
        this.container = container
    }
}